### ISSUE TITLE
How to get ALL group Members on large groups

### ISSUE THREAD

**Issue by wranger** [Adldap2/Adldap2#342]
Hi,

when I search a group, how could I get all group Members?

GroupA  (5 Members):
<code>
$group = $pro_adint->search()->groups()->where("samaccountname","=",'grp-all-admins")->first();
var_dump($group->member);
//Returns: array(5) { [0]=> string(76) "CN=USER1"} <- works perfect
</code>

GroupB (11.000Members)
`
$group = $pro_adint->search()->groups()->where("samaccountname","=",'grp-all-students")->first();
var_dump($group->member);
//Returns: NULL
`

When I do a var_dump($group) the ["member"] Attribute changed to ["member;range=0-4999"].

How to get the group members from large groups? And what happens if I do a save?
---

Here is my comment for a fix to the old adldap:
MS-AD return max 5000 elements during one request, so we have to split the request into several ... 
see https://msdn.microsoft.com/en-us/library/cc223242.aspx
 this very important for the member -attribute of a  group, we need a paging mechanism and change the attribute from "member" -> "memeber;range=0-1000" and increment it to invinit, the last element is "member;range=akt-val-*" the default value ist 1500, and max of 5000 -> this is hardcoded by MS :(

**Comment by stevebauman** [comment]
Have you tried retrieving the group, and then finding all members of the group with another paginated query?

```php
$group = $pro_adint->search()->groups()->where("samaccountname","=",'grp-all-students")->first();

$filter = "(memberof:1.2.840.113556.1.4.1941:={$group->getDn()})";

$results= $provider->search()->rawFilter($filter)->paginate();
```

**Comment by wranger** [comment]
Hi,

that works, but its very slow, 20sec and needs a lot of memory>512MB (PHP Fatal error:  Allowed memory size of 536870912 bytes exhausted).

----------------
The $group->addMembers() method dosen't need a $group->save(), the $group->setMembers() does!?

**Comment by stevebauman** [comment]
> The $group->addMembers() method dosen't need a $group->save(), the $group->setMembers() does!?

There is no method named `addMembers()`. There is a method named `addMember($member)` however.

I can understand the confusion between the two, and maybe this should be changed.

At the time, I thought it would be more fluent to simply call `addMember()` to add a single member to a group and be done with the operation, rather than `setMembers()` would be used with other `set` operations. For example:

```php
$user = $provider->search()->users()->find('Jane Doe');

$group = $provider->search()->groups()->find('Accounting');

$group->addMember($user);

return true;
```

```php
$john= $provider->search()->users()->find('John Doe');
$jane= $provider->search()->users()->find('Jane Doe');

$group = $provider->search()->groups()->find('Accounting');

// Calling setMembers actually removes **all** members
// and sets them to the specified array of members.
$group->setMembers([
    $john->getDn(),
    $jane->getDn(),
]);
```

**Comment by stevebauman** [comment]
Also in regards to your proposed solution:

> Here is my comment for a fix to the old adldap:
MS-AD return max 5000 elements during one request, so we have to split the request into several ...
see https://msdn.microsoft.com/en-us/library/cc223242.aspx
this very important for the member -attribute of a group, we need a paging mechanism and change the attribute from "member" -> "memeber;range=0-1000" and increment it to invinit, the last element is "member;range=akt-val-*" the default value ist 1500, and max of 5000 -> this is hardcoded by MS :(

Are you able to provide a PR for this? I don't have any groups in my AD that contain that many members.

I can mock results, but without actual tests as well, I'd be coding blind.

**Comment by stevebauman** [comment]
Actually, have you tried?:

```php
$group = Adldap::search()->groups()->select(['member;range=0-1'])->find('Accounting');
```

**Comment by wranger** [comment]
No I did not but that works as well.

**Comment by stevebauman** [comment]
Hi @biwerr, thanks for the fix! Are you able to submit it as a PR?

**Comment by stevebauman** [comment]
This has been added thanks to @biwerr. It will be released will full support in the upcoming `v7.0`.

https://github.com/Adldap2/Adldap2/blob/master/src/Models/Group.php#L180-L234

**Comment by totesz09** [comment]
I know It's an old issue, but may I ask some help? 

I have a code, see below:

```
$paginator = $this->provider->search()
->setDn($dn)
->select(['objectclass', 'member', 'name', 'samaccountname', 'objectGUID'])
->whereHas('objectclass')
->listing()
->paginate(1000)

$entries = $paginator->getResults();
```

In this case my $entries are NULL. The MaxValRange on MS LDAP server is configured to 1500 by default.
 
How can I use biwerr solution in this case? 

Can you add me an example please?
